crates/networking/p2p/rlpx/eth/blocks.rs

```diff
@@ -119,26 +127,48 @@ impl GetBlockHeaders {
         };
 
         for _ in 0..limit {
-            match storage.get_block_header(current_block as u64) {
-                Ok(Some(block_header)) => {
-                    headers.push(block_header);
-                    current_block += block_skip
-                }
-                Ok(None) => {
+            let block_header_opt = match get_block_header(storage, current_block) {
+                Ok(block_header) => block_header,
+                Err(err) => {
+                    error!(%err, block_ref=%current_block, "Error accessing DB while building header response for peer");
                     break;
                 }
-                // TODO(#1073)
-                // Research what we should do when an error is found in a P2P request.
-                Err(err) => {
-                    error!("Error accessing DB while building header response for peer: {err}");
-                    return vec![];
+            };
+            let Some(block_header) = block_header_opt else {
+                trace!(block_ref=%current_block, "Block header not found");
+                break;
+            };
+            headers.push(block_header);
+
+            // Update to next block to fetch
+            match current_block {
+                // We don't support fetching headers by hash, unless the hash is
+                // part of the canonical chain, so we break here.
+                // TODO: we could support fetching by hash in descending order,
+                // by fetching the parent of each block.
```

Comment:
According to https://github.com/ethereum/devp2p/blob/master/caps/eth.md#getblockheaders-0x03 , even when fetching by hash, it should be part of the canonical chain, so maybe we should adapt the code/comments to that?

Reply:
[b8e2005](https://github.com/lambdaclass/ethrex/pull/4576/commits/b8e20058898d04767176fcdfa5995933d0f7acaf)